File: src/ru/org/linux/spring/DelIPController.java
/*
 * Copyright 1998-2009 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.spring;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.Timestamp;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.servlet.ModelAndView;

import ru.org.linux.site.*;
import ru.org.linux.util.ServletParameterParser;

@Controller
public class DelIPController {
  @RequestMapping(value="/delip.jsp", method= RequestMethod.POST)
  public ModelAndView delIp(HttpServletRequest request) throws Exception {
    Map<String, Object> params = new HashMap<String, Object>();
    Map<Integer, String> deleted = new HashMap<Integer, String>();
    
    Template tmpl = Template.getTemplate(request);
    HttpSession session = request.getSession();
    
    if (!tmpl.isModeratorSession()) {
      throw new AccessViolationException("Not moderator");
    }

    Connection db = null;
    try {
      db = LorDataSource.getConnection();

      String reason = new ServletParameterParser(request).getString("reason");
      String ip = new ServletParameterParser(request).getString("ip");
      String time = new ServletParameterParser(request).getString("time");

      Calendar calendar = Calendar.getInstance();
      calendar.setTime(new Date());

      if ("hour".equals(time)) {
        calendar.add(Calendar.HOUR_OF_DAY, -1);
      } else if ("day".equals(time)) {
        calendar.add(Calendar.DAY_OF_MONTH, -1);
      } else if ("3day".equals(time)) {
        calendar.add(Calendar.DAY_OF_MONTH, -3);
      } else {
        throw new UserErrorException("Invalid count");
      }

      Timestamp ts = new Timestamp(calendar.getTimeInMillis());
      params.put("message", "Удаляем темы и сообщения после "+ts.toString()+" с IP "+ip+"<br>");

      db.setAutoCommit(false);
    
      User moderator = User.getUser(db, (String) session.getValue("nick"));
    
      PreparedStatement st = null;
      ResultSet rs = null;
      CommentDeleter deleter = null;
      
      try {
        // Delete IP topics
        PreparedStatement lock = db.prepareStatement("SELECT id FROM topics WHERE postip=?::inet AND not deleted AND postdate>? FOR UPDATE");
        PreparedStatement st1 = db.prepareStatement("UPDATE topics SET deleted='t',sticky='f' WHERE id=?");
        PreparedStatement st2 = db.prepareStatement("INSERT INTO del_info (msgid, delby, reason) values(?,?,?)");
        lock.setString(1, ip);
        lock.setTimestamp(2, ts);
        st2.setInt(2, moderator.getId());
        st2.setString(3, reason);
       
        int topicCounter = 0;
        ResultSet lockResult = lock.executeQuery(); // lock another delete on this row
        while (lockResult.next()) {
          int mid = lockResult.getInt("id");
          st1.setInt(1,mid);
          st2.setInt(1,mid);
          st1.executeUpdate();
          st2.executeUpdate();
          topicCounter++;
        }
        st1.close();
        st2.close();
        lockResult.close();
        lock.close(); 
        params.put("topics", topicCounter);
    
        // Delete user comments
        deleter = new CommentDeleter(db);
    
        st = db.prepareStatement("SELECT id FROM comments WHERE postip=?::inet AND not deleted AND postdate>? ORDER BY id DESC FOR update");
        st.setString(1,ip);
        st.setTimestamp(2, ts);

        rs = st.executeQuery();
    
        while (rs.next()) {
          int msgid = rs.getInt("id");

          if (!deleter.getReplys(msgid).isEmpty()) {
            deleted.put(msgid, "пропущен");
            continue;
          }

          deleted.put(msgid, deleter.deleteComment(msgid, reason, moderator, -20));
        }

        params.put("deleted", deleted);        
      } finally {
        if (deleter!=null) {
          deleter.close();
        }
      
        if (rs!=null) {
          rs.close();
        }
      
        if (st!=null) {
          st.close();
        }
      }
      
      db.commit();
      
      return new ModelAndView("delip", params);
    } finally {
      if (db!=null) {
        db.close();
      }
    }
  }
}

File: src/main/scala/ru/org/linux/spring/DelIPController.scala
/*
 * Copyright 1998-2024 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.spring

import com.typesafe.scalalogging.StrictLogging
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.servlet.ModelAndView
import ru.org.linux.auth.AuthUtil.ModeratorOnly
import ru.org.linux.auth.IPBlockDao
import ru.org.linux.comment.DeleteService
import ru.org.linux.search.SearchQueueSender
import ru.org.linux.user.UserErrorException

import java.sql.Timestamp
import java.time.Instant
import java.time.OffsetDateTime
import java.time.temporal.ChronoUnit
import scala.collection.mutable
import scala.jdk.CollectionConverters.{ListHasAsScala, MapHasAsJava}

@Controller
class DelIPController(searchQueueSender: SearchQueueSender, commentDeleteService: DeleteService,
                      ipBlockDao: IPBlockDao) extends StrictLogging {
  /**
   * Контроллер удаление топиков и сообщений по ip и времени
   *
   * @param reason причина удаления
   * @param ip     ip по которому удаляем
   * @param time   время за которое удаляем (hour, day, 3day, 5day)
   * @return возвращаем страничку с результатом выполнения
   */
  @RequestMapping(value = Array("/delip.jsp"), method = Array(RequestMethod.POST))
  def delIp(@RequestParam("reason") reason: String, @RequestParam("ip") ip: String, @RequestParam("time") time: String,
            @RequestParam(value = "ban_time", required = false) banTime: String,
            @RequestParam(value = "ban_mode", required = false) banMode: String): ModelAndView = ModeratorOnly { implicit currentUser =>
    val params = new mutable.HashMap[String, AnyRef]

    val delFrom = time match {
      case "hour" => Instant.now.minus(1, ChronoUnit.HOURS)
      case "day" => Instant.now.minus(1, ChronoUnit.DAYS)
      case "3day" => Instant.now.minus(3, ChronoUnit.DAYS)
      case "5day" => Instant.now.minus(5, ChronoUnit.DAYS)
      case _ => throw new UserErrorException("Invalid count")
    }

    val ts = new Timestamp(delFrom.toEpochMilli)

    params.put("message", s"Удаляем темы и сообщения после $ts с IP $ip<br>")

    val moderator = currentUser.user

    if (banTime != null) {
      val banTo = banTime match {
        case "hour" => Some(OffsetDateTime.now.plusHours(1))
        case "day" => Some(OffsetDateTime.now.plusDays(1))
        case "month" => Some(OffsetDateTime.now.plusMonths(1))
        case "3month" => Some(OffsetDateTime.now.plusMonths(3))
        case "6month" => Some(OffsetDateTime.now.plusMonths(6))
        case "unlim" => None
        case "remove" => Some(OffsetDateTime.now)
        case _ => throw new UserErrorException("Invalid count")
      }

      var allowPosting = false
      var captchaRequired = false

      banMode match {
        case "anonymous_and_captcha" =>
          allowPosting = true
          captchaRequired = true
        case "anonymous_only" =>
          allowPosting = true
          captchaRequired = false
        case _ =>
          allowPosting = false
          captchaRequired = false
      }

      ipBlockDao.blockIP(ip, moderator.getId, reason, banTo.orNull, allowPosting, captchaRequired)
    }

    val deleteResult = commentDeleteService.deleteByIPAddress(ip, ts, reason)

    params.put("topics", Int.box(deleteResult.getDeletedTopicIds.size))
    params.put("comments", Int.box(deleteResult.getDeletedCommentIds.size))
    params.put("skipped", deleteResult.getSkippedComments)

    for (topicId <- deleteResult.getDeletedTopicIds.asScala) {
      searchQueueSender.updateMessage(topicId, true)
    }

    searchQueueSender.updateComment(deleteResult.getDeletedCommentIds)

    logger.info("Deleted {} from {} by moderator {}: topic={}; comments={}", ip, time,
      currentUser.user.getNick, deleteResult.getDeletedTopicIds.size, deleteResult.getDeletedCommentIds.size)

    new ModelAndView("delip", params.asJava)
  }
}